presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -1000,19 +1020,33 @@ private void addPartitionsWithoutStatistics(String databaseName, String tableNam
         }
     }
 
+    private <V> V wrapMetastoreCallable(MetastoreContext metastoreContext, MetastoreCallable<V> callable)
+            throws Exception
+    {
+        if (!isHmsImpersonationEnabled) {
+            HiveMetastoreClient client = clientProvider.createMetastoreClient(null);
+            return callable.call(client, metastoreStats.get(client.getAddress()));
+        }
+        try (HiveMetastoreClient client = clientProvider.createMetastoreClient(null)) {
```

Comment:
Why do we need to keep this client open during the nested call?

How about

```
Optional<Token> delegationToken = Optiona.empty();

if(isHmsImpersonationEnabled){
  try (HiveMetastoreClient client = clientProvider.createMetastoreClient(null)) {
delegationToken = client.getMetastoreClient().get_delegation_token...
}
HiveMetastoreClient realClient = clientProvider.createMetastoreClient(token);
return callable.call(realClient, metastoreStats.get(realClient.getAddress()));

}


```

Reply:
Also let's move it all the way down, below all the usages

Reply:
in your code, if impersonation disabled, it still call createMetastoreClient for twice, we should do it only once if it is disabled.  I don't think we need to keep the nested client open, is there any concerns?